### ISSUE TITLE
Restore file size checking to HDF5DataStore

### ISSUE THREAD

**Issue by bieryAtFnal** [DUNE-DAQ/dfmodules#34]
We have a max_file_size_bytes configuration parameter for HDF5DataStore already included in the system, but it is not currently being used.  (We noticed some problems with file size checking near the end of the preparations for minidaqapp v1.0, so we disabled it.)

For dunedaq-v2.4.0, we would like to be able to specify maximum files sizes and have new files opened when older ones reach the size limit.  New files would have incrementing values for the "file index" in the filename.  For example, here are some possible filenames:

- `fake_minidaqapp_run000333_file0000_biery_20210211T165107.hdf5`
- `fake_minidaqapp_run000333_file0001_biery_20210211T165107.hdf5`
- `fake_minidaqapp_run000333_file0002_biery_20210211T165107.hdf5`

One question is how strictly this limit should be enforced.  A very strict enforcement may have some performance penalty.  (More on this in a later post.)

In the past, I have also advertised the possibility of specifying a maximum duration for a given raw data file.  This would allow us to switch files every half hour, for example.  A simple-minded implementation of this functionality that relies on an event interval that is << than the file duration would be straightforward, but questions may arise regarding what to do when there isn't any data for a while.  Would we need to start a watchdog thread that would cause the file to be closed even when data isn't flowing?  Questions like that make me think that it would be good to defer the file duration functionality for a little while.

Some of the reasons for wanting periodic file closing include being able to look at the data as a run progresses rather than having to wait until the end.  Also, at a teststand, If one of the files accidentally becomes corrupt, only a fraction of data taken in a long (e.g. over the weekend) run is lost.

**Comment by bieryAtFnal** [comment]
Some comments on implementation:

- as shown in the example filenames in the introduction to this Issue, I'm currently using a single timestamp in the "unique" part of the filename.  I did this so that we don't have to incur the time needed to fetch the time from the Linux system and format the time string for each file; we only have to do it once per run at Start time.  We can change this behavior so that each filename has a unique time, if needed, but it didn't seem necessary to me so far.
- I changed my mind several times about whether to update the DataStore interface to have a `write(list of data blocks)` method (like we have been talking about for a long time) or whether to have a `write(TriggerRecord)` method.  I ended up with the former simply because it needed fewer changes to the existing code.  if we decide that we want to add a `write(TriggerRecord)` method later, that should be very straightforward.  The pros and cons of the two approaches include A) making life easier for the calling code (`write(TriggerRecord)` is easy for callers) and B) burdening DataStore implementations with knowledge of TriggerRecord implementation details.

**Comment by bieryAtFnal** [comment]
Regarding the strictness of the file size limit...

In my initial tests, I noticed that the files being produced were about 3% larger than the size limit that I had specified.  I believe that this is due to the HDF5 file overhead.  Inside the code, I was simply counting the number of bytes in the TriggerRecordHeader and the Fragments, and I didn't take into account any possible overhead.  

Obviously, some of the options for dealing with this include:

1. adding code that looks at the actual file on disk to see how big it is.  That seems like it could be resource intensive, so we should think about whether that performance hit is worth it
2. keeping the simple behavior of counting bytes in TRHs and Fragments and just accepting that the files will be slightly larger than what was requested
3. adding a (configurable?) overhead factor in the code so that we attempt to take into account the file format overhead
4. have users take into account the overhead factor when they specify the size limit in the configuration.

For now, I'll stick with the simplest behavior (option 2), until we have an opportunity to discuss this.
